### ISSUE TITLE
ECMDERR Failed to execute "git ls-remote --tags --heads git://github.com/angular/bower-angular.git", exit code of #128

### ISSUE THREAD

**Issue by molyakov** [Hacklone/private-bower#102]
Hello,

When I use private bower registry it always trying to download packages using git
When I use registry on bower.herokuapp.com either https and git is used to download packages (not sure how bower decide what protocol to use)
This error I get when trying to install angular via private bower registry (configured in .bowerrc)
ECMDERR Failed to execute "git ls-remote --tags --heads git://github.com/angular/bower-angular.git", exit code of #128

But when I use bower.herokuapp.com registry (i.e. default) there is no any problem to install angular
bower angular#1.3.9                                  download https://github.com/angular/bower-angular/archive/v1.3.9.tar.gz

**Comment by Hacklone** [comment]
try running ```bower cache clean```. It kinda looks like the public bower registry also gives back git://github.com/angular/bower-angular.git I don't yet know how it goes for an archive :)

**Comment by ldlnxs** [comment]
Same issue here. Full error msg:

bower not-cached    git://localhost:6789/bootstrap#*
bower resolve       git://localhost:6789/bootstrap#*
bower ECMDERR       Failed to execute "git ls-remote --tags --heads git://localhost:6789/bootstrap", exit code of #128

Additional error details:
fatal: unable to connect to localhost:
localhost[0: ::1]: errno=Connection refused
localhost[1: 127.0.0.1]: errno=Connection refused
localhost[2: fe80::1%lo0]: errno=Undefined error: 0

**Comment by ldlnxs** [comment]
I was trying to examine the issue more but didn't come to any helpful conclusion. What I tried so far is checking to which extend the git configuration of the server might be the cause. To no avail up until now :(

**Comment by jliang6** [comment]
I have similar problem as follows:

bower ECMDERR       Failed to execute "git ls-remote --tags --heads git://localhost:6789/json3", exit code of #128

Additional error details:
fatal: unable to access 'https://localhost:6789/json3/': Unknown SSL protocol error in connection to localhost:6789

New Github key has been setup as well and that solved the problem of retrieving libraries from GitHub.  Now that JS libraries are available inside the local filesystem directory "gitRepoCache" (After the new Github key), however, my project still cannot receive the libraries because of the Exit code of 128 and Unknown SSL protocol error.

Please let me know if you guys have seen this or solve this before.

**Comment by ldlnxs** [comment]
Till running out of time, I came close to assume the issue is connected to 
+ restrictive firewall settings with port blocks
+ use of a proxy
+ inside a windows server environment
What I observed was that simple packages with a few levels of dependencies could be resolved quite nicely. From a certain point of package complexity on the thing hit a wall.

Unfortunately no one here has the time to track the thing down. We just stopped trying to use private-bower.

**Comment by zhanghan177** [comment]
`git config --global url."https://".insteadOf git://` solves my issue. 

https://github.com/bower/bower/issues/713

**Comment by jakehockey10** [comment]
@zhanghan177 your suggestion has solved my issue that was very similar to this.  Thank you!

**Comment by ant2400** [comment]
@zhanghan177 thanks a lot that worked perfectly!

**Comment by samuelarcher** [comment]
@zhanghan177 thanks -- worked for me as well.

**Comment by Matheo13** [comment]
@zhanghan177 thanks -- worked for me as well.

**Comment by mrmayfield** [comment]
@zhanghan177  Yup, works for me as well. Pretty certain I'm behind  a firewall that's shooting me down on this network.

**Comment by error500** [comment]
thanks

**Comment by deividfortuna** [comment]
perfect! thanks @zhanghan177

**Comment by jaw4ik** [comment]
thanks!!!

**Comment by harjolgoo** [comment]
@zhanghan177  that helps. Thanks.

**Comment by rhernandog** [comment]
@zhanghan177 That did it in my case, many thanks.

**Comment by SamanShafigh** [comment]
@zhanghan177 thanks

**Comment by suryateja619** [comment]
@zhanghan177 +1 It works pretty well.

**Comment by DarkSmile92** [comment]
@zhanghan177 thanks, solved it for me too!

**Comment by spiderOO7** [comment]
thanks work here too

**Comment by bulaluis** [comment]
@zhanghan177 thanks

**Comment by vnussolution** [comment]
It solved my issue. @zhanghan177 thanks

**Comment by Avlessi** [comment]
@zhanghan177 thanks you, sir!

**Comment by overmedia** [comment]
@zhanghan177 solved my issue

**Comment by sureshpaladugu2** [comment]
@zhanghan177 Thanks solved my issue

**Comment by vino4all** [comment]
@zhanghan177 thanks you, Saved my day.

**Comment by pedrolucasdsc** [comment]
Obrigado! (thanks you) @zhanghan177

**Comment by cesarokada** [comment]
thank you @zhanghan177 , you save me

**Comment by abellion** [comment]
@zhanghan177 thanks !!

**Comment by skorski** [comment]
Thank you @zhanghan177.  I had a feeling that was what it was and you saved me a few hours of guesswork.

**Comment by stephan1** [comment]
I got the same error . I solved it by reinstalling new version of git on windows 7 .

**Comment by pallaviMN** [comment]
@zhanghan177 , Can you please help in the below issue, i tried comment shared by you, but it didnt work for me 

bower invalid-meta  The "name" is recommended to be lowercase, can contain digits, dots, dashes
bower not-cached    git://github.com/shprink/ionic-native-transitions.git#*
bower resolve       git://github.com/shprink/ionic-native-transitions.git#*
bower cached        https://github.com/rajeshwarpatlolla/ionic-datepicker.git#0.9.0
bower validate      0.9.0 against https://github.com/rajeshwarpatlolla/ionic-datepicker.git#0.9.0
bower ECMDERR       Failed to execute "git ls-remote --tags --heads git://github.com/shprink/ionic-native-transitions.git", exit code of #128 fatal: repository 'https://github.com/shprink/ionic-native-transitionsgithub.com/shprink/ionic-native-transitions.git/' not found

Additional error details:
fatal: repository 'https://github.com/shprink/ionic-native-transitionsgithub.com/shprink/ionic-native-transitions.git/' not found

**Comment by zhanghan177** [comment]
@pallaviMN As far as I'm concerned, the method I mentioned only change "git://...." into "https://..."
However, in your case, if I try to load your "ionic-native-transitions" repo in my browser, I got no results. So it's likely that your project includes some no-longer-exist repos.

**Comment by metouole** [comment]
thank Zhanghan177 your solution also work well for me

**Comment by jeffloo8912** [comment]
thank Zhanghan177 very much

**Comment by uoc1691** [comment]
Another reason would be bower CLI was not able to communicate with git installed in the machine. Try adding the git to the PATH or opening the git terminal and then do the bower install.

**Comment by lightswitch05** [comment]
Interesting  quirk with this same issue. `bower install` works without trouble, it gets this error on `bower update`. The `git config` specified above appears to have no effect either way with `bower  update`.

**Comment by deepakkumar138** [comment]
why & how do we use : 
git config --global url."https://".insteadOf git://

**Comment by zhanghan177** [comment]
@deepakkumar138 https://stackoverflow.com/questions/1722807/git-convert-git-urls-to-http-urls

**Comment by chaosyuan** [comment]
@zhanghan177 thanks a lot

**Comment by Momar-NDIAYE** [comment]
thank you

**Comment by theunexpected1** [comment]
Thanks @zhanghan177, but funnily it didn't help me; it did show me where to look tho :)

Instead, [this](https://github.com/bower/bower/issues/1998#issuecomment-155812342) worked:
```git config --global url."https://github.com/".insteadOf git@github.com:```

**Comment by punnashivakrishna** [comment]
ECMDERR Failed to execute "git ls-remote --tags --heads https://github.com/angular/bower-angular.git", exit code of #128 fatal: unable to access 'https://github.com/angular/bower-angular.git/': error
 setting certificate verify locations:   CAfile: C:/Program Files/Git/mingw64/ssl/certs/ca-bundle.crt   CApath: none

**Comment by Chekote** [comment]
This is an authentication issue and is due to Bower trying to fetch via authenticated HTTPS or SSH.

If you have SSH keys setup so you do not need to enter a username and/or password when using SSH, then you will want to configure Git to use SSH by default:

`git config --global url.git@github.com:.insteadOf https://github.com/`

If on the other hand, you have HTTPS credentials stored in such a way that that method does not prompt you for username or password, then you will want to configure Git to use HTTPS by default:

`git config --global url.https://github.com/.insteadOf git@github.com:`

Your preference will depend entirely on which you have configured to automatically authenticate.

**Comment by thinkocapo** [comment]
i tried some of the git config commands in here and now I can't push to github anymore for my other projects. ugh. will research it:
```
Permission denied (publickey).
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
```

and this did not solve my issue but seems to have solved everyone else's:
`git config --global url."https://".insteadOf git://`

**Comment by pradeepta-patra** [comment]
Installing updated version of GIT solved my issue 
**[exec] bower requirejs#~2.1.11 ECMDERR Failed to execute "git ls-remote --tags --heads https://github.com/jrburke/requirejs-bower.git", ex
it code of #128 fatal: unable to access 'https://github.com/jrburke/requirejs-bower.git/': error:1407742E:SSL routines:SSL23_GET_SERVER_HELLO:tlsv1 alert protoc
ol version**

**Comment by Dectinc** [comment]
Another possible cause is that you are using incompatible versions of nss and curl, try updating these two modules.

For example, on CentOS, just run `yum upgrade -y nss curl`

**Comment by ivanzamoraarias** [comment]
@zhanghan177 not working at all -_-

**Comment by Chaithanya92** [comment]
git config --global url."git://".insteadOf https://

It could be due to being blocked from any access to bower installation. But this worked for me.

**Comment by frank198008** [comment]
if you use babun, try copy "C:\Program Files\Git\mingw64\ssl\certs\ca-bundle.crt" TO C:\Users\{UserName}\.babun\cygwin\etc\pki\tls\certs, that worked for me

**Comment by joquijada** [comment]
What solved the issue for me was these commands. My problem was that `gulp -b <app name>` was complaining that JQuery was not installed. After trial and error below is the combination of commands that did the trick for me. I did also do `bower cache clean`, not sure if that a difference as well,

1. `git config --global url."https://".insteadOf git://`
2. `bower install --force # The bower install messages were saying that some dependencies like jquery were skipped, and to issue the --force flag, so I gave it a try.` 
3. Then finally `gulp -b <app name>` ran successfully
